miniapps/common/particles_extras.cpp

```diff
@@ -0,0 +1,247 @@
+                                           const char *keys_)
+   : pset(particles), tail_size(tail_size_), x(x_), y(y_), w(w_), h(h_),
+     title(title_), keys(keys_)
+#ifdef MFEM_USE_MPI
+   ,comm(particles.GetComm())
+#endif // MFEM_USE_MPI
+{
+   sock.open(vishost, visport);
+   sock.precision(8);
+   newly_opened = true;
+
+   AddSegmentStart();
+}
+
+void ParticleTrajectories::AddSegmentStart()
+{
+   // Create a new mesh for all particle segments for this timestep
+   segment_meshes.emplace(segment_meshes.begin(), 1, pset.GetNParticles()*2,
+                          pset.GetNParticles(),
+                          0, pset.GetDim());
+
+   // Add segment start particle IDs
+   segment_ids.emplace(segment_ids.begin(), pset.GetIDs());
+
+   if (tail_size > 0 && static_cast<int>(segment_meshes.size()) > tail_size)
+   {
+      segment_meshes.pop_back();
+      segment_ids.pop_back();
+   }
+
+   // Add all particle starting vertices
+   for (int i = 0; i < pset.GetNParticles(); i++)
+   {
+      Vector pcoords;
+      pset.Coords().GetValues(i, pcoords);
+      segment_meshes.front().AddVertex(pcoords);
+   }
+}
+
+void ParticleTrajectories::SetSegmentEnd()
+{
+   const Array<unsigned int> &end_ids = pset.GetIDs();
+
+   // Add all endpoint vertices + segments for all particles
+   int num_start = segment_ids.front().Size();
+   for (int i = 0; i < num_start; i++)
+   {
+      // If this particle's initial position was set in AddSegmentStart,
+      // set the vertex to its now current location
+      int pidx = end_ids.Find(segment_ids.front()[i]);
+      if (pidx != -1)
+      {
+         Vector pcoords;
+         pset.Coords().GetValues(pidx, pcoords);
+         segment_meshes.front().AddVertex(pcoords);
+      }
+      else // Otherwise set its end vertex == start vertex
+      {
+         segment_meshes.front().AddVertex(segment_meshes.front().GetVertex(i));
+      }
+      segment_meshes.front().AddSegment(i, i+num_start);
+   }
+   segment_meshes.front().FinalizeMesh();
+
+}
+
+
+void ParticleTrajectories::Visualize()
+{
+   SetSegmentEnd();
+
+   // Create a mesh of all the trajectory segments
+   std::vector<Mesh*> all_meshes;
+   for (Mesh &m : segment_meshes)
+   {
+      all_meshes.push_back(&m);
+   }
+
+   Mesh trajectories(all_meshes.data(), all_meshes.size());
+
+   // Update socketstream
+#ifdef MFEM_USE_MPI
+   int myid, num_procs;
+   MPI_Comm_rank(comm, &myid);
+   MPI_Comm_size(comm, &num_procs);
+
+   sock << "parallel " << num_procs << " " << myid << "\n";
+#endif // MFEM_USE_MPI
+   sock << "mesh\n";
+   trajectories.Print(sock);
+
+
+   if (newly_opened)
+   {
+      sock << "window_title '" << title << "'\n"
+           << "window_geometry "
+           << x << " " << y << " " << w << " " << h << "\n";
+      if ( keys ) { sock << "keys " << keys << "\n"; }
+      else { sock << "keys maaAc\n"; }
+      sock << std::endl;
+      newly_opened = false;
+   }
```

Comment:
Replace by `VisualizeMesh()` from `(p)fem_extras`.

Reply:
I think that works for static vis and not when we need to repeatedly update the window

Reply:
no, should work